lib/NormalModule.js

```diff
@@ -136,17 +135,32 @@ class NormalModule extends Module {
 	}
 
 	createLoaderContext(resolver, options, compilation, fs) {
+		const addErrorFrom = (error) => {
+			if(!error.from) {
+				const currentLoaderObject = loaderContext.loaders && loaderContext.loaders.length ?
+					loaderContext.loaders[loaderContext.loaderIndex] :
+					null;
+				if(currentLoaderObject) {
+					const loaderName = currentLoaderObject.origin || currentLoaderObject.path;
+					error.from = /\//.test(loaderName) ?
+						path.relative(loaderContext.rootContext, loaderName) :
+						loaderName;
+				}
+			}
+		};
 		const loaderContext = {
 			version: 2,
 			emitWarning: (warning) => {
 				if(!(warning instanceof Error))
 					warning = new NonErrorEmittedError(warning);
+				addErrorFrom(warning);
 				this.warnings.push(new ModuleWarning(this, warning));
```

Comment:
Don't attach the `from` property to the existing Error object.

Add a constructor argument to `ModuleWarning`, etc. instead passing `from`.